File: IT/2/Multithreading/Esercizi/Java/Esercizio17.java
/* file Esercizio17.java */

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.Semaphore;

class Car implements Runnable {
	private Semaphore semaphore;

	public Car(Semaphore semaphore) {
		this.semaphore = semaphore;
	}

	@Override
	public void run() {
		try {
			Thread auto = Thread.currentThread();

			// andatura dell'auto compresa tra 45 e 90 km orari 
			int speed = ThreadLocalRandom.current().nextInt(45, 90+1);
			System.out.println("L'auto " + auto.getName() + " viaggia a " + speed + " km/h...");
			int pause = ThreadLocalRandom.current().nextInt(3, 6);
			auto.sleep(pause*1000);

			semaphore.acquire();

			System.out.println(auto.getName() + " attraversa l'incrocio.");
		}
		catch (InterruptedException e) {
			e.printStackTrace();
		}
		finally {
			semaphore.release();
		}
	}
}

public class Esercizio17 {
	public static void main(String [] args) {
		Semaphore semaphore = new Semaphore(3);

		for (int i = 0; i < 25; i++) {
			try {
				Thread car = new Thread(new Car(semaphore));
				car.start();
				int pause = ThreadLocalRandom.current().nextInt(0, 2);
				Thread.currentThread().sleep(pause*1000);
			}
			catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}
}

File: IT/2/Multithreading/Esercizi/Python/esercizio17.py
# file esercizio17.py

from __future__ import print_function
from multiprocessing import Process, current_process, Semaphore, Lock
from random import randint
from time import sleep

def car(semaphore, lock):
	auto = current_process().name

	# andatura dell'auto compresa tra 45 e 90 km orari
	speed = randint(45, 90)
	with lock:
		print("L'auto {} viaggia a {} km/h...".format(auto, speed))

	pause = randint(3, 5)
	sleep(pause)

	c = semaphore.acquire()
	# print "{}, c={}".format(auto, c)
	# if c == True:
	with lock:
		print("{} attraversa l'incrocio.".format(auto))

	semaphore.release()
	# else:
	#	print "L'auto {} si ferma in coda".format(auto)

if __name__ == "__main__":
	semaphore = Semaphore(3)
	lock = Lock()

	auto = [Process(target=car, args=(semaphore, lock,)) for i in range(25)]

	for p in auto:
		p.start()
		pause = randint(0, 1)
		sleep(pause)

	for p in auto:
		p.join()

